tracer startup logs in troubleshooting page (#7901)
DataDog/documentation: content/en/tracing/troubleshooting/_index.md

title: APM Troubleshooting
kind: documentation
further_reading:
- link: "/tracing/troubleshooting/tracer_startup_logs/"
  tag: "Documentation"
  text: "Datadog tracer startup logs"
- link: "/tracing/troubleshooting/tracer_debug_logs/"
  tag: "Documentation"
  text: "Datadog tracer debug logs"
- link: "/tracing/troubleshooting/agent_apm_metrics/"
  tag: "Documentation"
  text: "APM metrics sent by the Datadog Agent"
---

When experiencing unexpected behavior with Datadog APM, there are a few common issues you can look for before reaching out to [Datadog support][1]:

> 1. **Make sure the Agent has APM enabled**:

Reviewer: I think it still makes sense to keep some mention of the agent in this doc since it is required under most circumstances.  If we're removing this in favor of pointing to other documentation we should likely reference that doc here in my opinion.

> All Datadog tracing libraries emit logs during startup by default to reflect the configurations applied, as well as any errors encountered during startup.  These startup logs are available from the minimum versions of the tracer indicated in the below table:

Author: I reordered this section a bit so all the text is above the table - tried to incorporate this

| Language | Version |
|----------|---------|
| Java    |  0.59 (once available)  |
| .NET | 1.18.2+  |
| PHP | 0.47.0+  |
| Go | 1.26.0 (once available)  |
| NodeJS | 0.23.0 (once available)  |
| Python | 0.41 (once available)  |
| Ruby | 0.38 (once available)  |
| C++ | 1.1.6 (once available)  |

> If your tracer version includes these [startup logs][2], they are the recommended place to begin troubleshooting.

Reviewer: None of the tracers will require any additional logging config similar to enabling debug mode correct?  This doc makes it sound like that initialize logging will happen by default  regardless of debug mode but maybe we should say so explicitly.

## Tracer debug logs

In order to capture full details on the Datadog tracer, debug mode can be enabled on your tracer via the environment variable `DATADOG_TRACE_DEBUG`.  This may be done for your own investigation or recommended by Datadog support for triage purposes. However, it is not recommended to have debug mode always enabled due to logging overhead.

These logs can surface instrumentation errors or integration-specific errors.  To see further details on how to enable and capture these debug logs, please see the [debug mode troubleshooting page][3].

## APM rate limits

### Max events per second limit

If you encounter the following error message in your Agent logs, your application(s) are emitting more than the default 200 trace events per second allowed by APM.

```
Max events per second reached (current=300.00/s, max=200.00/s). Some events are now being dropped (sample rate=0.54). Consider adjusting event sampling rates.

```

> To increase the APM rate limit for the Agent, configure the `max_events_per_second` attribute within the Agent's configuration file (underneath the `apm_config:` section). For containerized deployments (Docker, Kubernetes, etc.), use the `DD_APM_MAX_EPS` environment variable.

Reviewer: The use of the word "limit" is totally accurate and justified here  and in the following section but it does seem to contradict some of the market around "tracing without limits".  In my view we should probably tone down the "without limits" rhetoric instead of adapting here but perhaps we should avoid the word "limit" where possible.

**Note**: Increasing the APM rate limit could result in increased costs for App Analytics.

### Max connection limit

If you encounter the following error message in your Agent logs, the default APM connection limit of 2000 has been exceeded:

```
ERROR | (pkg/trace/logutil/throttled.go:38 in log) | http.Server: http: Accept error: request has been rate-limited; retrying in 80ms
```

To increase the APM connection limit for the Agent, configure the `connection_limit` attribute within the Agent's configuration file (underneath the `apm_config:` section). For containerized deployments (Docker, Kubernetes, etc.), use the `DD_APM_CONNECTION_LIMIT` environment variable.

## Troubleshooting data requested by Datadog Support

When you open a [support ticket][1], our support team may ask for a combination of the below information:

> 1. How you are confirming the issue (links to a trace, screenshots, etc) and what you expect to see

Reviewer: I think bold formatting on all these numbered headings would help users focus their eyes on key points.

    This allows us to confirm errors and attempt to reproduce your issues within our testing environments.

1. [Tracer Startup Logs](#tracer-startup-logs)

    Startup logs are a great way to spot misconfiguration of the tracer, or the inability for the tracer to communicate with the Datadog Agent.  By comparing the configuration that the tracer sees to the one set within the application or container, we can identify any areas where a setting is not being properly applied.

1. [Tracer Debug Logs](#tracer-debug-logs)

>     Tracer Debug logs go one step deeper than startup logs, and will help us to identify if integrations are instrumenting properly in a manner that we aren't able to necessarily check until traffic flows through the application.

Reviewer: We might want to add something like this:
"It can be extremely useful for viewing the contents of spans created by the tracer and can surface an error if there is a connection issue when attempting to send spans to the agent.  Tracer debug logs are typically the most informative source of information and a reliable tool for confirming nuanced behavior of the tracer."

> 1. An [agent flare][4] from your agent in [debug mode][5]

Reviewer: I would probably link directly to this doc for sending a flare instead of the more general Agent Troubleshooting page.
https://docs-staging.datadoghq.com/andrew.ardito/tracerstartuplogs/agent/troubleshooting/send_a_flare/?tab=agentv6v7

Reviewer: I might be a tiny bit more specific in this heading and use something like:
"An agent flare (snapshot of logs and configs) that captures a representative log sample from a time period when traces where sent and the agent was in debug mode"

Author: > I would probably link directly to this doc for sending a flare instead of the more general Agent Troubleshooting page.
> https://docs-staging.datadoghq.com/andrew.ardito/tracerstartuplogs/agent/troubleshooting/send_a_flare/?tab=agentv6v7

[4] is exactly that, just without the tab but it automatically will input the tab when navigating there

>     Agent flares allow us to see what is happening within the Datadog Agent, if any traces are being rejected or malformed within the Agent.  This will not help if traces are not reaching the Agent, but does help us identify the source of an issue, as well as identifying any metric discrepancies.

Reviewer: I wonder if this might be a time to mention the `TRACE` level logging as that does help us resolve issues with deeper logging than `DEBUG`.  I suspect we don't advertise `TRACE` more often because it's very verbose logging but this might be the right circumstance to mention it.  Perhaps it's discussing with other PMs and/or the agent team.

1. A description of your environment

    Knowing how your application is deployed helps us identify likely issues for tracer-agent communication problems or misconfigurations.  For difficult issues, we may ask to a see a Kubernetes manifest or an ECS task definition, for example.

1. Any automatic or [custom instrumentation][6], along with any configurations

    Custom instrumentation can be a very powerful tool, but also can have unintentional side effects on your trace visualizations within Datadog, so we ask about this to rule it out as a suspect.  Additionally, asking for your automatic instrumentation and configuration allows us to confirm if this matches what we are seeing in both tracer startup and debug logs to look for discrepancies.

> 1. Versions of languages, frameworks, the Datadog Agent and Tracing Library being used

Reviewer: I think convention would be to have a comma after "Datadog Agent" right before the "and".

Reviewer: a nit pick i know 😿

During startup, all Datadog tracing libraries past the versions listed below emit logs that reflect the configurations applied in a JSON object, as well as any errors encountered, including if the Agent can be reached in languages where this is possible.  If your tracer version includes these [startup logs][2], start your troubleshooting there.

| Language | Version |
|----------|---------|
| Java    |  0.59 (once available)  |
| .NET | 1.18.2+  |
| PHP | 0.47.0+  |
| Go | 1.26.0 (once available)  |
| NodeJS | 0.23.0+  |
| Python | 0.41 (once available)  |
| Ruby | 0.38 (once available)  |
| C++ | 1.1.6 (once available)  |

## Tracer debug logs

To capture full details on the Datadog tracer, enable debug mode on your tracer by using the `DATADOG_TRACE_DEBUG` environment variable. You might enable it for your own investigation or because  Datadog support recommended it for triage purposes. However, don't leave debug mode always enabled because of the logging overhead it introduces.

These logs can surface instrumentation errors or integration-specific errors.  For details on enabling and capturing these debug logs, see the [debug mode troubleshooting page][3].

## APM rate limits

> Within Datadog Agent logs, if you see error messages about rate limits or max events per second, these limits can be adjusted by following  [these instructions][4].  For any questions before changing limits, consult with our [support team][1].

Reviewer 2: ```suggestion
Within Datadog Agent logs, if you see error messages about rate limits or max events per second, you can these limits by following [these instructions][4].  If you have questions, before you change the limits, consult with our [support team][1].
```

1. **Any automatic or [custom instrumentation][7], along with any configurations**

    Custom instrumentation can be a very powerful tool, but also can have unintentional side effects on your trace visualizations within Datadog, so we ask about this to rule it out as a suspect.  Additionally, asking for your automatic instrumentation and configuration allows us to confirm if this matches what we are seeing in both tracer startup and debug logs.

1. **Versions of languages, frameworks, the Datadog Agent, and Tracing Library being used**

    Knowing what versions are being used allows us to ensure integrations are supported, to check for known issues, or to recommend a tracer or language version upgrade if it will address the problem.

1. **Confirm Agent configurations, including if APM is enabled**

    While APM is enabled for Agent 6+, in containerized environments there is an additional configuration step for [non local traffic][8] and can be the solution to traces not being received.

>     You can check this by running the command `netstat -van | grep 8126` on your Agent host. If you don't see an entry, this means you should check if your Agent is running or the configuration of your `datadog.yaml` file. Instructions can be found on the [Agent Configuration page][8].

Reviewer 2: ```suggestion
    You can check this by running the command `netstat -van | grep 8126` on your Agent host. If you don't see an entry, this means you should check that your Agent is running or check the configuration in your `datadog.yaml` file. Instructions can be found on the [Agent Configuration page][8].
```
